Unicode UTF-8 Identifiers in USD Proposal

erslavin opened the pull request:
### Description of Proposal

Currently, identifiers in USD are limited to ASCII character sets.  As such, Unicode based languages cannot use their native character sets to name prims and properties.  This proposal discusses what it would take to add Unicode support to core USD identifiers.

### Supporting Materials

https://github.com/PixarAnimationStudios/USD/pull/2120

### Contributing

<!--
Please review the  [Contributing](https://graphics.pixar.com/usd/release/contributing_to_usd.html) page in the
documentation for the Supplemental Terms that apply to this repository.
Place an X in the box when you have reviewed and agree to the Supplemental Terms.
-->
- [X] I agree to and accept the [Supplemental Terms](https://graphics.pixar.com/usd/release/contributing_supplemental.html).

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+# Staged Rollout and Adoption
+
+Since there are potentally large code bases in the current ecosystem that rely on USD, this proposal currently lands changes behind a `TfEnvSetting` (`TF_UTF8_IDENTIFIERS`) such that Pixar and other VFX/animation stakeholders are not forced to adopt UTF-8 identifier support within their code bases right away.  This environment setting defaults to `false` such that current ASCII validation rules a ...
```
It might be worth considering a `CMAKE` flag instead of the environment variable given that you use the `Ar` rollout as analogy.  The environment variable might be harder to deprecate?

It's also worth noting that it's probably important to get the sorting algorithm right since it's hard to change given the concerns about diff-ing files.  To ensure that layers are easily interchanged and that UTF-8 supported widely, I'd argue that the USD library should be opinionated about when the feature is considered a preview and when it's considered a first class feature.  Perhaps `TF_PREVIEW_UTF8_IDENTIFIERS` or (if you accept the `CMAKE` suggestion), `PXR_PREVIEW_UTF8_IDENTIFIERS`.

erslavin replied:
The `CMAKE` flag is an interesting option, and would alleviate their concerns about a runtime environment switch.  I can put this as an option under this section and state that while we implemented it in the PR as a runtime switch, the `CMAKE` flag provides some additional flexibility and eliminates the runtime cost of evaluating the flag.

erslavin replied:
Added additional note at end of section to discuss a `CMAKE` flag approach

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+- UTF-8
+- UTF-16 / UTF-32
+
+In evaluating options to support, raw code points rely largely on the support of code pages and individual system locales.  UTF-x encodings are already populate for standardized interchange of Unicode text and in particular UTF-8 encoding is widespread within the community.  UTF-8 is a variable length encoding that is backward compatible with existing ASCII encodings, meaning that current usd fil ...
```
The `UTF-8 Everywhere` doc was a very compelling read, and I'd argue for its inclusion earlier in the document to seed context for the rest of the proposal.

erslavin replied:
Added a reference to this doc early on after the Unicode standard itself.

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+Though support may vary, both the C++ standard and PEP 3131 have common definitions for what denotes an identifier and the recommendation is to follow this convention in USD, particularly since it must conform to these rules to enable successful compilation of code-based schemas.  One downside of this approach is that strings starting with digits (or indeed, all digit strings) could not be consid ...
+
+Restricting prim names in this way, however, is largerly unnecessary given they are never used by compilers generating schema code.  In fact, it is entirely possible prim names could be any valid Unicode string (e.g., parentheses, spaces, emojis, etc.).  While this proposal does not specify the possible characters that could potentially make up a valid prim name, it does suggest a way to separate ...
```
It's interesting and important to note that prim names don't necessarily need to be restricted to identifiers in the same way that properties do. Mentioning emojis and other characters as possible prim identifiers might distract from the current proposal. I might argue for keeping this higher level and noting that there might be opportunities to extend the character set for prims in future proposals?

erslavin replied:
I can see your point.  On the other hand, this gets people thinking about the possibilities of native character sets in the names.  I think there is an argument to be made for both restricting and expanding the set of characters allowed.  The emojis might be a little distracting though :)

erslavin replied:
Removed emojis and added a **NOTE** to the beginning to illustrate the discussion nature of the paragraph

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+- `xid_start_range_class`: A vector of pairs (`low`, `high`) of code points that fall in the `XID_Start` character class
+- `xid_continue_range_class`: A vector of pairs (`low`, `high`) of code points that fall in the `XID_Continue` character class
+
+This preprocessed file would then be compiled with the rest of the code and be available for fast lookup at runtime.  Preprocessing is necessary becuase:
```
Spelling error.  "because"

erslavin replied:
fixed

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+This preprocessed file would then be compiled with the rest of the code and be available for fast lookup at runtime.  Preprocessing is necessary becuase:
+
+- The Unicode database is not formatted in such a way to provide fast run-time lookups of character classes in a memory efficient manner
+- The Unicode database is large, and not all information is required to perform the necessary processing
```
Have we profiled a memory impact from the Unicode database?

erslavin replied:
We have not in any recordable sense, though we did take a look at the library size between the versions that did and did not have the compile time tables to support the lookups.  The database itself is preprocessed, so the memory overhead is that used to store the various maps we generate and compile into the library.

erslavin replied:
I didn't add additional descriptive text here since we haven't really documented the increase in size

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+One complication is the use of flex and bison to generate the state machines and code necessary to lex and parse these sequences.  While we would like to limit changes such that only flex is required to regenerate the code and state machine, it is necessary to also regenerate the parser generator through bison.  Unfortunately, different versions of flex and bison generate code in slightly differe ...
+
+Note that all of the above relies on the source data being UTF-8 encoded.  This should not be a problem for legacy ASCII based `usda` files or code and should maintain backward compatibility with no changes required (as long as extended character sets are not used).  Moving forward, any USD content that wants to take advantage of Unicode support must encode content in UTF-8.  This includes both f ...
```
Would it be prudent to consider a usda version bump?

erslavin replied:
Since we've never seen a bump, I don't know the right answer to that question.  It makes logical sense and perhaps we should include the suggestion in the proposal (especially in a diffing scenario where the ordering really does differ in ASCII properties between a version 1.0 file and a version 2.0 file).

nvmkuruc replied:
Yeah.  I'm thinking something like `#usda 1.0` files were strictly ASCII but `#usda 1.1` files are strictly UTF-8.

erslavin replied:
added additional content to discuss bumping the version

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+Note that all of the above relies on the source data being UTF-8 encoded.  This should not be a problem for legacy ASCII based `usda` files or code and should maintain backward compatibility with no changes required (as long as extended character sets are not used).  Moving forward, any USD content that wants to take advantage of Unicode support must encode content in UTF-8.  This includes both f ...
+
+Since may tools in the USD ecosystem are Python based, the translation layer between Python and C++ should also be considered.  While in most cases this works out of the box, errors can occur if strings at the Python layer are not UTF-8 encoded before being passed to the C++ layer.  Particular attention must be paid to the way current Python code reads in file content (e.g., schema generation, et ...
```
"Since may tools" should probably be "Since many tools"

erslavin replied:
fixed

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+Note that all of the above relies on the source data being UTF-8 encoded.  This should not be a problem for legacy ASCII based `usda` files or code and should maintain backward compatibility with no changes required (as long as extended character sets are not used).  Moving forward, any USD content that wants to take advantage of Unicode support must encode content in UTF-8.  This includes both f ...
+
+Since may tools in the USD ecosystem are Python based, the translation layer between Python and C++ should also be considered.  While in most cases this works out of the box, errors can occur if strings at the Python layer are not UTF-8 encoded before being passed to the C++ layer.  Particular attention must be paid to the way current Python code reads in file content (e.g., schema generation, et ...
```
Do you mean `boost::python` instead of `boost::bind`?

erslavin replied:
fixed

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+  - `TfStringEndsWith`
+- String concatenation - Unicode normalization forms are not closed under string concatenation; specifically at least the area where the strings are joined need to be considered for composition
+  - `TfStringJoin`
+
```
It might be worth having an opinion about C++20 `std::u8string`.

erslavin replied:
We can certainly point to the docs around its proposal.  I don't think anything in this proposal works differently in a `char8_t` world (except maybe you could have overloads that are specifically ASCII vs. explicitly UTF-8 and keep the current paths for `char` - which puts the burden on the developer to choose which to use).  The major advantage would be the compiler encoding string literals in code with the proper encoding (UTF-8) rather than relying on compiler defaults (which may be UTF-8 and may not).

nvmkuruc replied:
Makes sense.  Any opinions or interesting API observations due to C++20/23 can be captured in other documents.

erslavin replied:
added an additional sentence discussing `std::u8string`

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+- String concatenation - Unicode normalization forms are not closed under string concatenation; specifically at least the area where the strings are joined need to be considered for composition
+  - `TfStringJoin`
+
+## Identifier Collation
```
I think I'd title this `Identifier Sorting` and note that UCA is the standard way to collate strings for user presentation, but other sorting algorithms can more performantly provide valid orderings.

My reasoning is that APIs like ICU do make a distinction between their collation APIs and their compare / sort APIs and we shouldn't make them synonymous.

erslavin replied:
I can change it - to be honest the definitions are so similar that different explanations in different places use one to define the other.  I'll go check how ICU defines the difference semantically.

nvmkuruc replied:
That would be helpful.  I drew a clear line in my head when experimenting with python.  `locale.strcoll` is one specific way to generate a valid sorting of strings in python.

erslavin replied:
Changed name of header and added an additional couple of sentences describing there could be other approaches than the ones enumerated here.

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+The sub-strings `myString002` and `myString` can be formed for comparison purposes to retrieve the result.  Note that the `x` here is not included in the substring because only digits can form the varying length portion of substring extraction.  This results in `myString` < `myString002`.  The algorithm continues to compare `x` and `ß`.  Since `x` is code point less than `ß` and this is the only  ...
+
+For an optimistic ASCII algorithm to be used in practice, it would be required to explicit lay out the context-sensitive rules being used by the existing dictionary ordering algorithm.  These rules would then have to be adjusted to allow for what happens when non-ASCII characters "break" the context - are they ignored?  do they contribute a new context?  does the algorithm terminate at that point ...
+
```
This is a great explanation of a really hard problem.

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+For an optimistic ASCII algorithm to be used in practice, it would be required to explicit lay out the context-sensitive rules being used by the existing dictionary ordering algorithm.  These rules would then have to be adjusted to allow for what happens when non-ASCII characters "break" the context - are they ignored?  do they contribute a new context?  does the algorithm terminate at that point ...
+
+## Unicode Collation Algorithm
```
This is a really good section, but given that "optimistic ASCII" is going to be the sorting algorithm, I wonder if it should be its own referenced file to capture the work and information while keeping the main proposal focused on the changes to USD Core.

erslavin replied:
Captured in the overall comment below.

asluk commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+Aaron Luk  
+Matthew Kuruc  
+Edward Slavin  
```
if you're comfortable with it, feel free to put yourself as primary author, @erslavin -- I feel you've done the most heavy lifting with this effort.

nvmkuruc replied:
Agreed!

asluk replied:
@erslavin -- sounds like in general NV is using the term AECO (architecture, engineering, construction, and operations) rather than AEC-- let's reflect that here as well. Thanks!

erslavin replied:
made the adjustment to AECO

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+Autodesk is one of the primary stakeholders currently asking for UTF-8 identifiers, as Revit and 3ds max already support UTF-8 element names.  While Maya does not currently support UTF-8, we anticipate that it will eventually also be in demand, and the VFX/animation community will then be pushed to follow suit in their internal pipelines.  With the opt-in approach outlined above, this proposal ai ...
+
+We share concern over UTF-8 identifiers leaking into pipelines that do not yet support them, and stand ready to offer guidance and support to update those use cases.  As most VFX/animation facilities source their content in-house and/or within non-UTF-8 ecosystems, we do not expect UTF-8 identifiers to appear unexpectedly in such pipelines unless intentionally ingested.  IN any case, the `TfEnvSe ...
```
Should "IN any case" be "In any case"?

erslavin replied:
fixed

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+# An Overview of Identifiers in USD
+
+Identifier syntax is currently enforced in USD by several different mechanisms.  The primary mechanism is the inlined function `TfIsValidIdentifier`, which states that "an identifier is valid if it follows the C/Python identifier convention; that is, it must be at least one charaacter long, must start with a letter or underscore, and must contain only letters, underscores, and numerals".  However ...
```
Should charaacter be character?

erslavin replied:
fixed

nvmkuruc replied:
This still seems to be there in the latest unless I'm missing an update.

erslavin replied:
hmm, you are right, must have been a Ctrl+Z malfunction, fixed

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,435 @@
+
+It is also necessary to define what we mean by *identifier*.  The current implementation mixes two syntactically similar but semantically different concepts:
+
+- **Prim names:** The name of a USD prim, differentiating one sibling prim from another and forming the hierarchical path from which a prim can be referenced
```
I think this is a good point and worth capturing, but for focus and given that the proposal isn't as this time recommending that prim names extend beyond identifiers, it might be worth condensing or moving to a sibling document.

erslavin replied:
Captured in the comment below.

asluk reviewed (commented):
Overall looks good! I'll give it a more thorough review tomorrow, but in the meantime it will be good to have a separate "out of scope" section to clarify that while AEC content often has whitespace, parentheses, etc., display names ARE the right approach there, i.e., like in C++ and Python, it does make sense to draw some reasonable limitations around identifier syntax in USD, even if it means not supporting everything that AECO DCCs support. Thanks!

nvmkuruc reviewed (commented):
I left some comments in line and have two main notes about topics that might be able to be streamlined.

I think the proposal does not need to spend a lot of time addressing whether or not prim names should be restricted to identifiers. It's definitely notable, but there's compelling reasons to keep it restricted too. Rather than expand the section to include that debate, it's probably better to keep it focused on what's needed for UTF-8 Identifier Support.

Secondly, I think the proposal should acknowledge collation / UCA but ultimately not spend a lot of time on it.  The documentation you've provided on collation is really good and should be preserved perhaps in a sibling markdown file, but ultimately if the core is going to use optimistic ASCII for its internal ordering / sorting, it's secondary to the main proposal.

erslavin commented:
> I left some comments in line and have two main notes about topics that might be able to be streamlined.
> 
> I think the proposal does not need to spend a lot of time addressing whether or not prim names should be restricted to identifiers. It's definitely notable, but there's compelling reasons to keep it restricted too. Rather than expand the section to include that debate, it's probably better to keep it focused on what's needed for UTF-8 Identifier Support.
> 
> Secondly, I think the proposal should acknowledge collation / UCA but ultimately not spend a lot of time on it. The documentation you've provided on collation is really good and should be preserved perhaps in a sibling markdown file, but ultimately if the core is going to use optimistic ASCII for its internal ordering / sorting, it's secondary to the main proposal.

This content was from the original proposal where we did actually modify the prim name rules slightly and spent a fair amount of time implementing UCA.  I'd like to keep them in there, but it looks like I need to make a clearer statement somewhere that there are no changes to prim name rules planned (it's important, because we did do the work to separate out the validation, one just falls back to the other when there are no changes to the rules).  As far as UCA, we never finalized that optimistic ASCII was the way to go (and I have some doubts in its current state).  That said, I agree it likely isn't the final choice, so perhaps I preface the whole collation section with the fact that the subsections are different proposals, we still need to decide which route to take (with tradeoffs mentioned in the different sections) and that because of the performance hit and / or effort needed to cache that data, we likely wouldn't go with UCA?

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

- Addressed review comments

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

- Removed transparency in diagram for dark backgrounds

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,443 @@
+- String concatenation - Unicode normalization forms are not closed under string concatenation; specifically at least the area where the strings are joined need to be considered for composition
+  - `TfStringJoin`
+
+It may be interesting to consider the impact of C++ 20's `std::u8string` type for the methods listed above.  Overloads of these methods that take a `std::u8string` parameter would provide very clear semantics to a developer about the treatment of string content.  While the use of `std:u8string` would really only affect the copmiler encoding of string literals, the burden is then placed on the dev ...
```
copmiler  => compiler

erslavin replied:
Proofreading is hard :/  fixed! (and thanks!)

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,443 @@
+
+For lexer-based validation, a new lexer and parser are introduced to specifically capture new identifier rules for both the path lexer and the text file format lexer.  Although it is possible to change the existing lexer / parser in such a way to support UTF-8 encodings and new identifier rules, we thought it may introduce subtle errors across the ecosystem when using the modified lexer / parser  ...
+
+The path lexer is used as the first step in validating prim path strings.  The text file format lexer is used to read and validate the content of text-based usd files (i.e., `.usda`).  Both require that a sequence of characters be recognized as a logical token (and parsed in context) before being validated.  The rules that dictate what lexer sequences are acceptable for identifiers need to be upd ...
```
staht => that

erslavin replied:
fixed

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,443 @@
+    - [Core Identifier Validation](#core-identifier-validation)
+    - [The Path and Text File Format Lexer / Parser](#the-path-and-text-file-format-lexer--parser)
+    - [Case Mapping and other String Operations](#case-mapping-and-other-string-operations)
+    - [Identifier Collation](#identifier-collation)
```
Should this be Identifier Sorting?

erslavin replied:
fixed

nvmkuruc commented on proposals/tf_utf8_identifiers/README.md:
```diff
@@ -0,0 +1,443 @@
+
+In evaluating options to support, raw code points rely largely on the support of code pages and individual system locales.  UTF-x encodings are already populate for standardized interchange of Unicode text and in particular UTF-8 encoding is widespread within the community.  UTF-8 is a variable length encoding that is backward compatible with existing ASCII encodings, meaning that current usd fil ...
+
+Building USD is already a complex endevor.  While there are existing libraries that do provide APIs for working with Unicode character sequences, like [ICU](https://icu.unicode.org/), using these libraries would necessitate the acquisition and distribution of this library with the core USD libraries.  Efforts are already underway to eliminate dependencies (e.g., Boost), so we felt that a small en ...
```
Should this be "the choice of _sorting_ algorithm", "understanding the _sorting_ algorithm", and a link to "Identifier Sorting"

erslavin replied:
fixed

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

erslavin committed:
Unicode UTF-8 Identifiers in USD Proposal

erslavin committed:
Updated to reflect chosen sort algorithm of Standard ASCII Dictionary Ordering

erslavin committed:
Fixed Standard ASCII Dictionary Sort Ordering Header

erslavin committed:
Fixed typos

erslavin committed:
Fixed header for UCA

erslavin committed:
Added additional notes to make clear which sorting algorithm was selected and which was not

erslavin committed:
Unicode UTF-8 Identifiers

- Updated proposal to reflect changed default setting to
  TF_UTF8_IDENTIFIERS runtime switch
- Updated to make clear the distinction (and validation rules used)
  between prim names and Identifiers
